forage-android/src/main/java/com/joinforage/forage/android/vault/BasisTheoryPinSubmitter.kt

```diff
@@ -66,51 +67,57 @@ internal class BasisTheoryPinSubmitter(
         // The caller should have already performed the error checks.
         // We add these error checks as a safeguard, just in case.
-        if (toVaultErrorOrNull(vaultResponse) != null ||
+        if (vaultResponse.isFailure ||
+            toVaultErrorOrNull(vaultResponse) != null ||
             toForageErrorOrNull(vaultResponse) != null
         ) {
             return null
         }
 
+        // note: Result.toString() wraps the actual response as
+        // "Success(<actual-value-here>)"
         return ForageApiResponse.Success(vaultResponse.getOrNull().toString())
     }
 
     override fun toForageErrorOrNull(vaultResponse: BasisTheoryResponse): ForageApiResponse.Failure? {
-        // is a Basis Theory error and doesn't have to do with Forage
-        if (vaultResponse.isFailure) return null
-
         return try {
-            // if the response is a ForageApiError, then this block
-            // should not throw
-            val forageResponse = vaultResponse.getOrNull().toString()
-            val forageApiError = ForageApiError.ForageApiErrorMapper.from(forageResponse)
-            val error = forageApiError.errors[0]
-            ForageApiResponse.Failure.fromError(
-                ForageError(400, error.code, error.message)
+            val matchedStatusCode = getBasisTheoryExceptionStatusCode(vaultResponse)!!
+            val basisTheoryExceptionBody = getBasisTheoryExceptionBody(vaultResponse)!!
+
+            val forageApiError = ForageApiError.ForageApiErrorMapper.from(basisTheoryExceptionBody)
+            val firstError = forageApiError.errors[0]
+
+            return ForageApiResponse.Failure.fromError(
+                ForageError(matchedStatusCode, firstError.code, firstError.message)
             )
-        } catch (_: JSONException) {
-            // if we throw when trying to extract the ForageApiError,
+        } catch (_: Exception) {
+            // if we throw (likely a NullPointerException) when trying to extract the ForageApiError,
             // that means the response is not a ForageApiError
             null
         }
     }
 
     override fun toVaultErrorOrNull(vaultResponse: BasisTheoryResponse): ForageApiResponse.Failure? {
-        // for basis theory, Success responses mean Basis Theory did not
-        // mess up. So, it will be a Forage Success or Forage Error
-        // but it won't be a Basis Theory Error
         if (vaultResponse.isSuccess) return null
 
-        // if the result is a Failure, we need to fail gracefully and
-        // return a ForageApiError
-        return UnknownErrorApiResponse
+        try {
+            // try to parse as an error from Basis Theory, indicated by the presence
+            // of a "proxy_error" in the raw response body.
+            val basisTheoryExceptionBody = getBasisTheoryExceptionBody(vaultResponse)
+            if (basisTheoryExceptionBody!!.contains("proxy_error")) {
```

Comment:
Maybe we should add some logging around this? I'm not sure I'm 100% comfortable with all this string parsing it seems like it could be a recipe for uncaught exceptions.

Reply:
Well put, I'm on board.